File: src/config/env.ts
import { z } from 'zod';

/**
 * All runtime configuration for Lites is sourced from environment variables
 * and validated once, at startup, through this schema.
 *
 * Fail fast: if the environment is misconfigured, the process should refuse
 * to start rather than run in an unknown state. This matters especially in
 * containers, where a bad deploy should fail health checks immediately
 * instead of serving traffic in a broken configuration.
 */
const envSchema = z.object({
  NODE_ENV: z.enum(['development', 'test', 'production']).default('development'),

  // Network
  HOST: z.string().default('0.0.0.0'),
  PORT: z.coerce.number().int().positive().default(3000),

  // Logging
  LOG_LEVEL: z.enum(['fatal', 'error', 'warn', 'info', 'debug', 'trace', 'silent']).default('info'),

  // Graceful shutdown
  SHUTDOWN_TIMEOUT_MS: z.coerce.number().int().positive().default(10_000),

  // Reserved for later phases (providers, cache, routing). Kept optional and
  // unused for now so Phase 0 has no hidden dependency on external services.
  ANTHROPIC_API_KEY: z.string().optional(),
  OPENAI_API_KEY: z.string().optional(),
  REDIS_URL: z.string().optional(),
});

export type Env = z.infer<typeof envSchema>;

let cachedEnv: Env | undefined;

/**
 * Parses and validates `process.env` on first call, then returns the cached
 * result. Throws synchronously (crashing the process) if required variables
 * are missing or malformed — this is intentional for Phase 0.
 */
export function loadEnv(source: NodeJS.ProcessEnv = process.env): Env {
  if (cachedEnv) {
    return cachedEnv;
  }

  const parsed = envSchema.safeParse(source);

  if (!parsed.success) {
    const issues = parsed.error.issues
      .map((issue) => `  - ${issue.path.join('.')}: ${issue.message}`)
      .join('\n');
    throw new Error(`Invalid environment configuration:\n${issues}`);
  }

  cachedEnv = parsed.data;
  return cachedEnv;
}

/** Test-only escape hatch to reset the memoized config between test cases. */
export function __resetEnvCacheForTests(): void {
  cachedEnv = undefined;
}

File: app/config/env.py
from typing import Literal
from pydantic_settings import BaseSettings, SettingsConfigDict
from typing import Optional

class Env(BaseSettings):
    NODE_ENV: Literal["development", "test", "production"] = "development"
    HOST: str = "0.0.0.0"
    PORT: int = 3000
    LOG_LEVEL: Literal["fatal", "error", "warn", "info", "debug", "trace", "silent"] = "info"
    SHUTDOWN_TIMEOUT_MS: int = 10000
    
    # Optimization Thresholds
    MIN_TOKENS_FOR_OPTIMIZATION: int = 50
    MAX_TOKENS_FOR_OPTIMIZATION: int = 128000
    
    # AI Optimization configuration
    AI_OPTIMIZE_THRESHOLD: int = 500
    OPENAI_API_KEY: str = ""
    GEMINI_API_KEY: str = ""
    
    # Redis configuration
    REDIS_URL: Optional[str] = None
    CACHE_TTL_SECONDS: int = 3600
    
    # Authentication
    LITES_API_KEY: Optional[str] = None
    
    # Security / CORS
    FRONTEND_URL: str = "http://localhost:5173"
    
    # Semantic Cache configuration
    SEMANTIC_CACHE_THRESHOLD: float = 0.95

    model_config = SettingsConfigDict(
        env_file=".env",
        env_file_encoding="utf-8",
        extra="ignore"
    )

env = Env()
